### PR TITLE
Pallet ibc documentation

### LINKED ISSUE #1374: Documentation for pallet ibc
**Motivation**

Readme: This should contain a full technical documentation for pallet ibc.
Write comments for code where needed.

### PR DESCRIPTION
**Pull request by Wizdave97**
## Issue
Closes #1374

### CODE REVIEW

File: code/parachain/frame/ibc/README.md [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+	type Currency = Balances;
+	const INDEXING_PREFIX: &'static [u8] = b"ibc/";
+	const CONNECTION_PREFIX: &'static [u8] = b"ibc/";
+	const CHILD_TRIE_KEY: &'static [u8] = b"ibc/";
+	const LIGHT_CLIENT_PROTOCOL: pallet_ibc::LightClientProtocol = pallet_ibc::LightClientProtocol::Grandpa; // Finality protocol this chain will be using
+	type ExpectedBlockTime = ExpectedBlockTime; // Expected block time in milliseconds
+	type MultiCurrency = Assets; // Add a module that implements the MultiCurrency trait
+	type AccountIdConversion = ibc_primitives::IbcAccount;
+	type AssetRegistry = AssetsRegistry; // Add a module that implements the required trait
+	type CurrencyFactory = CurrencyFactory; // Add a module that impplements the required trait;
```
Comment (KaiserKarel) [inline code/parachain/frame/ibc/README.md:37]: nit: the doc comment here simply explains rust. I wouldn't mind if it was removed, or elaborated upon.

File: code/parachain/frame/ibc/README.md [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+
+- **PacketReceipt:** An opaque value committed to storage after a message receive packet has been successfully processed
+
+- **PacketAcknowledgement:** A sha256 hash of packet acknowledgement committed to runtime storage by modules
+
+- **NextSequenceSend:** A u64 value representing the next packet sequence to be sent on a channel
+
+- **NextSequenceReceive:** A u64 value representing the next packet sequence to be received on a channel
+
+- **NextSequenceAck:** A u64 value representing the next acknowledgement sequence to be received on a channel
```
Comment (KaiserKarel) [inline code/parachain/frame/ibc/README.md:62]:
```suggestion
- **ClientState:** This represents a connected chain's light client parameters, required for header verification.

- **ConsensusState:** This represents the timestamp and state root of a connected chain's light client at a particular block height, 
  it is extracted from the block header on successful processing of a ClientUpdate message, its major purpose is for proof verification.

- **PacketCommitment:** The sha256 hash of packet data, sequence,  timeout timestamp and timeout height committed to the runtime storage.

- **PacketReceipt:** An opaque value committed to storage after a message receive packet has been successfully processed.

- **PacketAcknowledgement:** A sha256 hash of packet acknowledgement committed to runtime storage by modules.

- **NextSequenceSend:** A u64 value representing the next packet sequence to be sent on a channel.

- **NextSequenceReceive:** A u64 value representing the next packet sequence to be received on a channel.

- **NextSequenceAck:** A u64 value representing the next acknowledgement sequence to be received on a channel.
```

Comment (KaiserKarel) [inline code/parachain/frame/ibc/README.md:62]:
```suggestion
- **ClientState:** This represents a connected chain's light client parameters, required for header verification.

- **ConsensusState:** This represents the timestamp and state root of a connected chain's light client at a particular block height, 
  it is extracted from the block header on successful processing of a ClientUpdate message, its major purpose is for proof verification.

- **PacketCommitment:** The sha256 hash of packet data, sequence,  timeout timestamp and timeout height committed to the runtime storage.

- **PacketReceipt:** An opaque value committed to storage after a message receive packet has been successfully processed.

- **PacketAcknowledgement:** A sha256 hash of packet acknowledgement committed to runtime storage by modules.

- **NextSequenceSend:** A u64 value representing the next packet sequence to be sent on a channel.

- **NextSequenceReceive:** A u64 value representing the next packet sequence to be received on a channel.

- **NextSequenceAck:** A u64 value representing the next acknowledgement sequence to be received on a channel.
```

Comment (KaiserKarel) [inline code/parachain/frame/ibc/README.md:62]:
```suggestion
- **ClientState:** This represents a connected chain's light client parameters, required for header verification.

- **ConsensusState:** This represents the timestamp and state root of a connected chain's light client at a particular block height, 
  it is extracted from the block header on successful processing of a ClientUpdate message, its major purpose is for proof verification.

- **PacketCommitment:** The sha256 hash of packet data, sequence,  timeout timestamp and timeout height committed to the runtime storage.

- **PacketReceipt:** An opaque value committed to storage after a message receive packet has been successfully processed.

- **PacketAcknowledgement:** A sha256 hash of packet acknowledgement committed to runtime storage by modules.

- **NextSequenceSend:** A u64 value representing the next packet sequence to be sent on a channel.

- **NextSequenceReceive:** A u64 value representing the next packet sequence to be received on a channel.

- **NextSequenceAck:** A u64 value representing the next acknowledgement sequence to be received on a channel.
```

**Review by KaiserKarel — APPROVED** [review]
Really like these docs overal

**Comment by KaiserKarel** [comment]
I think it would be good to apply the fixes and squash this into a single commit. You can instruct mergify to squash, or do it yourself before merging.

File: code/parachain/frame/ibc/README.md [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+- `transfer` - This initiates an ics20 token transfer from the caller to an account on a connected chain via the ICS20 application
+- `set_params` - Sets parameters that determine whether token transfer or receipt is allowed in ICS20
+- `upgrade_client` - Sets the new consensus state and client state for client upgrades to be executed on connected chains
+
+### Adding Ibc to a substrate runtime
+
+Implementing the ibc config trait for a substrate runtime
+```rust
+parameter_types! {
+	pub const ExpectedBlockTime: u64 = 6000;
```
Comment (seunlanlege) [inline code/parachain/frame/ibc/README.md:21]:
```suggestion
	pub const ExpectedBlockTime: u64 = 12000;
```

File: code/parachain/frame/ibc/README.md [hunk 4]
```diff
@@ -0,0 +1,525 @@
...
+
+Pallet IBC is a thin wrapper around [`ibc-rs`](/code/centauri/ibc) that satisfies the runtime requirements for communicating over the IBC protocol from a substrate runtime.
+
+- [`Config`](/code/parachain/frame/ibc/src/lib.rs#L204)
+- [`Call`](/code/parachain/frame/ibc/src/lib.rs#L533)
+- [`Pallet`](/code/parachain/frame/ibc/src/lib.rs#L268)
+
+### Dispatchable functions
+
+- `deliver` - Receives a batch of ibc messages and processes them
```
Comment (seunlanlege) [inline code/parachain/frame/ibc/README.md:11]:
```suggestion
- `deliver` - Receives a batch of ibc transactions and executes them in the same order as they were recieved
```

File: code/parachain/frame/ibc/README.md [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+- `transfer` - This initiates an ics20 token transfer from the caller to an account on a connected chain via the ICS20 application
```
Comment (seunlanlege) [inline code/parachain/frame/ibc/README.md:12]:
```suggestion
- `transfer` - This initiates an ics20 token transfer from the caller to an account on a connected chain via the ICS20 protocol
```

File: code/parachain/frame/ibc/README.md [hunk 6]
```diff
@@ -0,0 +1,525 @@
...
+	pub const RelayChainId: light_client_commomn::RelayChain = light_client_commomn::RelayChain::Rococo;
+}
+
+impl pallet_ibc::Config for Runtime {
+	type TimeProvider = Timestamp;
+	type Event = Event;
+	type Currency = Balances;
+	const INDEXING_PREFIX: &'static [u8] = b"ibc/";
+	const CONNECTION_PREFIX: &'static [u8] = b"ibc/";
+	const CHILD_TRIE_KEY: &'static [u8] = b"ibc/";
```
Comment (seunlanlege) [inline code/parachain/frame/ibc/README.md:31]: I wonder if it makes sense to refactor pallet ibc itself to just use one param? and use them for all 3

**Review by seunlanlege — CHANGES_REQUESTED** [review]
High quality docs overall

File: code/parachain/frame/ibc/README.md (hunk 6 above)
Comment (Wizdave97) [reply code/parachain/frame/ibc/README.md:31]: Yeah that makes sense.

**Pull request merged**